packages/replayio/src/utils/recordings/upload/uploadRecording.ts

```diff
@@ -198,54 +194,115 @@ async function uploadRecordingFileInParts({
-  );
+  const { size: totalSize } = await stat(recordingPath);
+  const abortController = new AbortController();
+
+  const partsUploads = partLinks.map((url: string, index: number) => {
+    return uploadQueue.add(async () => {
+      try {
+        return retryWithLinearBackoff(
+          async () => {
+            const partNumber = index + 1;
+            const start = index * chunkSize;
+            const end = Math.min(start + chunkSize, totalSize) - 1; // -1 because end is inclusive
+
+            debug("Uploading part %o", {
+              partNumber,
+              start,
+              end,
+              totalSize,
+              chunkSize,
+            });
+            return uploadPart(
+              { url, recordingPath, start, end, size: end - start + 1 },
+              abortController.signal
+            );
+          },
+          (error: any) => {
+            debug(`Failed to upload part ${index + 1}. Will be retried: %o`, error);
+            if (error.code === "ENOENT") {
+              throw error;
+            }
+          }
+        );
+      } catch (error) {
+        abortController.abort();
+        throw error;
+      }
+    });
+  });
 
-  return results;
+  return Promise.all(partsUploads);
 }
 
 async function uploadPart(
-  link: string,
-  partMeta: { recordingPath: string; start: number; end: number },
-  size: number
+  {
+    url,
+    recordingPath,
+    start,
+    end,
+    size,
+  }: {
+    url: string;
+    recordingPath: string;
+    start: number;
+    end: number;
+    size: number;
+  },
+  abortSignal: AbortSignal
 ): Promise<string> {
-  return new Promise((resolve, reject) => {
-    const worker = new Worker(join(__dirname, "./uploadWorker.js"));
-
-    worker.on("message", resolve);
-    worker.on("error", reject);
-    worker.on("exit", code => {
-      if (code !== 0) {
-        reject(new Error(`Worker stopped with exit code ${code}`));
-      }
-    });
+  debug("Uploading chunk %o", { recordingPath, size, start, end });
+  const stream = createReadStream(recordingPath, { start, end });
+  const response = await uploadRecordingReadStream(stream, { url, size }, abortSignal);
 
-    worker.postMessage({ link, partMeta, size, logPath: debugLogPath });
-  });
+  const etag = response.headers.get("etag");
+  assert(etag, "Etag has to be returned in the response headers");
+  debug("Etag received %o", { etag, recordingPath, size, start, end });
+
+  return etag;
+}
+
+async function uploadRecordingReadStream(
+  stream: ReadStream,
+  { url, size }: { url: string; size: number },
+  abortSignal?: AbortSignal
+) {
+  abortSignal?.throwIfAborted();
+
+  const streamError = createDeferred<never>();
+  const closeStream = () => stream.close();
+  stream.on("error", streamError.reject);
+
+  try {
+    const response = await Promise.race([
+      fetch(url, {
+        headers: {
+          "Content-Length": size.toString(),
+          "User-Agent": getUserAgent(),
+          Connection: "keep-alive",
```

Comment:
The previous version of the code was only using this for the multipart upload - I think it can safely be used at all times here. On top of that, the previous version of the code was creating a custom HTTP agent. It was only setting `keepAlive: true` option on it but that's already the default:
https://github.com/nodejs/node/blob/ee4fa77624f17c1958dac47fddbe7d9513b9dee8/lib/_http_agent.js#L557

Reply:
I'm not sure what node version we want to support but if I remember correctly this default was changed recently (Node 16/18 maybe?).

Reply:
Oh, goooood catch. Node 18 doesn't have this `keepAlive` there. I guess, I'll have to replicate the Node 20 behavior by creating a custom agent here (one that we'll be able to reuse across fetches)